packages/@lwc/ssr-compiler/src/compile-js/stylesheet-scope-token.ts

```diff
@@ -0,0 +1,56 @@
+import { is } from 'estree-toolkit';
+import { generateScopeTokens } from '@lwc/template-compiler';
+import { builders as b } from 'estree-toolkit/dist/builders';
+import { esTemplate } from '../estemplate';
+import type { BlockStatement, ExportNamedDeclaration, Program, VariableDeclaration } from 'estree';
+
+function generateStylesheetScopeToken(filename: string) {
+    // FIXME: we should be getting the namespace/name from the config options,
+    // since these actually come from the component filename, not the template filename.
+    const split = filename.split('/');
+    const namespace = split.at(-3)!;
+    const baseName = split.at(-1)!;
+
+    const componentName = baseName.replace(/\.[^.]+$/, '');
+    const {
+        // FIXME: handle legacy scope token for older API versions
+        scopeToken,
+    } = generateScopeTokens(filename, namespace, componentName);
+
+    return scopeToken;
+}
+
+const bStylesheetTokenDeclaration = esTemplate<VariableDeclaration>`
+    const stylesheetScopeToken = '${is.literal}';
+`;
+
+const bAdditionalDeclarations = [
+    esTemplate<VariableDeclaration>`
+        const hasScopedStylesheets = defaultScopedStylesheets && defaultScopedStylesheets.length > 0;
+    `,
+    esTemplate<ExportNamedDeclaration>`
+        const stylesheetScopeTokenClass = hasScopedStylesheets ? \` class="\${stylesheetScopeToken}"\` : '';
+    `,
+    esTemplate<ExportNamedDeclaration>`
+        const stylesheetScopeTokenHostClass = hasScopedStylesheets ? \` class="\${stylesheetScopeToken}-host"\` : '';
+    `,
+    esTemplate<ExportNamedDeclaration>`
+        const stylesheetScopeTokenClassPrefix = hasScopedStylesheets ? (stylesheetScopeToken + ' ') : '';
+    `,
+];
```

Comment:
There is probably a better way to do this, but I like the `esTemplate` utility and did not find a more succinct way to represent this.